modules/dbparser/tests/test_parsers.c

```diff
@@ -74,56 +56,72 @@ _invoke_parser(gboolean (*parser)(guint8 *str, gint *len, const gchar *param, gp
+    {.str = "foo{}", .param = NULL, .expected_string = "foo", .expected_result = TRUE},
+    {.str = "foo[]", .param = NULL, .expected_string = "foo", .expected_result = TRUE},
+    {.str = "foo", .param = "X", .expected_string = "foo", .expected_result = TRUE},
+    {.str = "foo=bar", .param = "=", .expected_string = "foo=bar", .expected_result = TRUE},
+    {.str = "", .param = NULL, .expected_string = NULL, .expected_result = FALSE},
+  };
+
+  return cr_make_param_array(ParserTestParam, parser_params,
+                             sizeof(parser_params) / sizeof(parser_params[0]));
 }
 
-static void
-assert_parser_failure(gboolean (*parser)(guint8 *str, gint *len, const gchar *param, gpointer state,
-                                         RParserMatch *match), const gchar *str, gpointer param, gpointer state)
+ParameterizedTest(ParserTestParam *param, parser, test_string_parser)
 {
+  gchar *result_string = NULL;
   gboolean result;
 
-  result = _invoke_parser(parser, str, param, state, NULL);
-  assert_false(result, "Mismatching parser result");
+  result = _invoke_parser(r_parser_string, param->str, param->param, NULL, &result_string);
+  if (param->expected_result == TRUE)
+    {
+      cr_assert(result, "Mismatching parser result");
+      cr_assert_str_eq(result_string, param->expected_string, "Mismatching parser result");
+      g_free(result_string);
+    }
+  else
+    {
+      cr_assert_not(result, "Mismatching parser result");
+    }
 }
 
-static void
-test_string_parser_without_parameter_parses_a_word(void)
+typedef struct _parser_test_qstring_param
 {
-  assert_parser_success(r_parser_string, "foo", NULL, NULL, "foo");
-  assert_parser_success(r_parser_string, "foo bar", NULL, NULL, "foo");
-  assert_parser_success(r_parser_string, "foo123 bar", NULL, NULL, "foo123");
-  assert_parser_success(r_parser_string, "foo{}", NULL, NULL, "foo");
-  assert_parser_success(r_parser_string, "foo[]", NULL, NULL, "foo");
-  assert_parser_failure(r_parser_string, "", NULL, NULL);
-}
+  gchar *str;
+  gchar *quotes;
+  gchar *expected_string;
+} ParserQStringTestParam;
 
-static void
-test_string_parser_with_additional_end_characters(void)
+ParameterizedTestParameters(parser, test_qstring_parser)
 {
-  assert_parser_success(r_parser_string, "foo", "X", NULL, "foo");
-  assert_parser_success(r_parser_string, "foo=bar", "=", NULL, "foo=bar");
-}
+  static ParserQStringTestParam parser_params[] =
+  {
+    {.str = "'foo'", .quotes = "''", .expected_string = "foo"},
+    {.str = "\"foo\"", .quotes = "\"\"", .expected_string = "foo"},
+    {.str = "{foo}", .quotes = "{}", .expected_string = "foo"},
+  };
 
-void
-test_string(void)
-{
-  PARSER_TESTCASE(test_string_parser_without_parameter_parses_a_word);
-  PARSER_TESTCASE(test_string_parser_with_additional_end_characters);
+  return cr_make_param_array(ParserQStringTestParam, parser_params,
+                             sizeof(parser_params) / sizeof(parser_params[0]));
```

Comment:
Can I suggest https://developer.gnome.org/glib/stable/glib-Standard-Macros.html#G-N-ELEMENTS:CAPS?

Reply:
I added the G_N_ELEMENTS macro.